File: efni/kennsluefni/frumtolur/code/pollard-rho.py
from math import gcd
from random import randrange

def miller_rabin(n, k):
    if n%2 == 0:
        return n == 2
    if n <= 3:
        return n == 3
    s = 0
    d = n - 1
    t = [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41]
    for _ in range(k): t.append(randrange(n - 2) + 1)
    while d%2 == 0:
        d //= 2
        s += 1
    for a in t:
        if a > n - 2: continue
        x = pow(a, d, n);
        if x == 1 or x == n - 1: continue
        e = True
        for i in range(s - 1):
            x = (x*x)%n
            if x == n - 1:
                e = False
                break
        if e: return False;
    return True;
def rho(n):
    s = [2, 3, 4, 5, 7, 11, 13, 1031]
    a = 1
    while True:
        for x in s:
            y = x
            d = 1
            while d == 1:
                x = (x*x + a)%n
                y = (y*y + a)%n
                y = (y*y + a)%n
                d = gcd(abs(x - y), n)
            if d != n: return d
        a += 1

def pollard_rho(n):
    a = []
    s = []
    p = [2, 3, 5, 7, 11, 13];
    for e in p:
        while n%e == 0:
            n //= e
            a.append(e)
    if n == 1: return a
    s.append(n)
    while (len(s) > 0):
        k = s[-1]
        del(s[-1])
        if miller_rabin(k, 10):
            a.append(k)
        else:
            r = rho(k);
            s.append(r)
            s.append(k//r)
    return a

print(pollard_rho(int(input())))

File: efni/kennsluefni/frumtolur/code/pollard_rho.java
import java.util.Scanner;
import java.util.Arrays;
import java.math.BigInteger;

public class pollard_rho
{
    public static void main(String[] args)
    {
        Scanner s = new Scanner(System.in);
        long[] r = pollard_rho(s.nextLong());
        for (int i = 0; i < r.length; i++) System.out.print(r[i] + " ");
        System.out.println();
    }

    static long gcd(long a, long b) { return b == 0 ? a : gcd(b, a%b); }
    static long bigprod(long x, long y, long m)
    {
        BigInteger r = BigInteger.valueOf(x);
        r = r.multiply(BigInteger.valueOf(y));
        r = r.mod(BigInteger.valueOf(m));
        return r.longValue();
    }
    static long bigsquare(long x, long a, long m)
    {
        BigInteger r = BigInteger.valueOf(bigprod(x, x, m));
        r = r.add(BigInteger.valueOf(a));
        r = r.mod(BigInteger.valueOf(m));
        return r.longValue();
    }
    static long modpow(long x, long n, long m)
    {
        long r;
        for (r = 1; n > 0; n = n/2, x = bigsquare(x, 0, m)) if (n%2 == 1) r = bigprod(r, x, m);
        return r; 
    }
    static boolean miller_rabin(long n)
    {
        if (n%2 == 0) return n == 2;
        if (n <= 3) return n == 3;
        long a, x, i, s = 0, d = n - 1;
        long[] t = {2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37};
        while (d%2 == 0) { d /= 2; s++; }
        for (int k = 0; k < t.length; k++) if (t[k] <= n - 2)
        {
            a = t[k]; x = modpow(a, d, n);
            if (x == 1 || x == n - 1) continue;
            for (i = 0; i < s - 1; i++) if ((x = bigsquare(x, 0, n)) == n - 1) break;
            if (i == s - 1) return false;
        }
        return true;
    }
    static long rho(long n)
    {
        long[] s = {2, 3, 4, 5, 7, 11, 13, 1031};
        long a, x, y, d;
        for (a = 1;; a++) for (int i = 0; i < 8; i++)
        {
            x = y = s[i]; d = 1;
            while (d == 1)
            {
                x = bigsquare(x, a, n);
                y = bigsquare(y, a, n);
                y = bigsquare(y, a, n);
                d = gcd(Math.abs(x - y), n);
            }
            if (d != n) return d;
        }
    }
    static long[] pollard_rho(long n)
    {
        int ss = 0, aa = 0;
        long[] s = new long[200], a = new long[200], p = {2, 3, 5, 7, 11, 13};
        for (int i = 0; i < p.length; i++) while (n%p[i] == 0) { n /= p[i]; a[aa++] = p[i]; }
        if (n != 1) s[ss++] = n;
        while (ss > 0)
        {
            long k = s[--ss];
            if (miller_rabin(k)) a[aa++] = k;
            else { long r = rho(k); s[ss++] = r; s[ss++] = k/r; }
        }
        return Arrays.copyOf(a, aa);
    }
}
